File: packages/std/src/predicates.ts
/**
 * A0 stdlib: boolean/predicate helpers
 * eq, contains, not, and, or
 */
import { isTruthy } from "@a0/core";
import type { StdlibFn, A0Record, A0Value } from "@a0/core";

function deepEqual(a: A0Value, b: A0Value): boolean {
  if (a === b) return true;

  if (a === null || b === null) return a === b;

  if (Array.isArray(a) || Array.isArray(b)) {
    if (!Array.isArray(a) || !Array.isArray(b)) return false;
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (!deepEqual(a[i] ?? null, b[i] ?? null)) return false;
    }
    return true;
  }

  if (typeof a === "object" && typeof b === "object") {
    const aRec = a as A0Record;
    const bRec = b as A0Record;
    const aKeys = Object.keys(aRec);
    const bKeys = Object.keys(bRec);
    if (aKeys.length !== bKeys.length) return false;
    for (const key of aKeys) {
      if (!Object.prototype.hasOwnProperty.call(bRec, key)) return false;
      if (!deepEqual(aRec[key] ?? null, bRec[key] ?? null)) return false;
    }
    return true;
  }

  return false;
}

/**
 * eq { a: <value>, b: <value> } -> boolean
 * Deep structural equality.
 */
export const eqFn: StdlibFn = {
  name: "eq",
  execute(args: A0Record): A0Value {
    const a = args["a"] ?? null;
    const b = args["b"] ?? null;
    return deepEqual(a, b);
  },
};

/**
 * contains { in: <string|list|record>, value: <value> } -> boolean
 * - string: substring check (value must be string)
 * - list: deep element membership via JSON.stringify
 * - record: key existence (value must be string)
 */
export const containsFn: StdlibFn = {
  name: "contains",
  execute(args: A0Record): A0Value {
    const input = args["in"] ?? null;
    const value = args["value"] ?? null;

    // string: substring check
    if (typeof input === "string") {
      if (typeof value !== "string") return false;
      return input.includes(value);
    }

    // list: deep element membership
    if (Array.isArray(input)) {
      return input.some((el) => deepEqual(el ?? null, value));
    }

    // record: key existence (value must be string)
    if (input !== null && typeof input === "object") {
      if (typeof value !== "string") return false;
      return Object.prototype.hasOwnProperty.call(input, value);
    }

    return false;
  },
};

/**
 * not { in: <value> } -> boolean
 * Boolean negation with A0 truthiness coercion.
 */
export const notFn: StdlibFn = {
  name: "not",
  execute(args: A0Record): A0Value {
    const input = args["in"] ?? null;
    return !isTruthy(input);
  },
};

/**
 * and { a: <value>, b: <value> } -> boolean
 * Logical AND with A0 truthiness coercion.
 */
export const andFn: StdlibFn = {
  name: "and",
  execute(args: A0Record): A0Value {
    const a = args["a"] ?? null;
    const b = args["b"] ?? null;
    return isTruthy(a) && isTruthy(b);
  },
};

/**
 * or { a: <value>, b: <value> } -> boolean
 * Logical OR with A0 truthiness coercion.
 */
export const orFn: StdlibFn = {
  name: "or",
  execute(args: A0Record): A0Value {
    const a = args["a"] ?? null;
    const b = args["b"] ?? null;
    return isTruthy(a) || isTruthy(b);
  },
};

/**
 * coalesce { in: val, default: fallback } -> value
 * Returns `in` if not null, else `default`. Strictly null-checking (NOT truthiness).
 */
export const coalesceFn: StdlibFn = {
  name: "coalesce",
  execute(args: A0Record): A0Value {
    const input = args["in"] ?? null;
    const fallback = args["default"] ?? null;
    return input !== null ? input : fallback;
  },
};

/**
 * typeof { in: val } -> str
 * Returns the A0 type name: "null", "boolean", "number", "string", "list", "record".
 */
export const typeofFn: StdlibFn = {
  name: "typeof",
  execute(args: A0Record): A0Value {
    const input = args["in"] ?? null;
    if (input === null) return "null";
    if (typeof input === "boolean") return "boolean";
    if (typeof input === "number") return "number";
    if (typeof input === "string") return "string";
    if (Array.isArray(input)) return "list";
    return "record";
  },
};

File: go/pkg/stdlib/predicates.go
package stdlib

import (
	"github.com/thomasrohde/agent0/go/pkg/evaluator"
)

// contains { in: string|list|record, value: any } → bool
func stdlibContains(args *evaluator.A0Record) (evaluator.A0Value, error) {
	input, _ := args.Get("in")
	value, _ := args.Get("value")
	if input == nil {
		input = evaluator.NewNull()
	}
	if value == nil {
		value = evaluator.NewNull()
	}

	switch in_ := input.(type) {
	case evaluator.A0String:
		// Substring check
		valStr, ok := value.(evaluator.A0String)
		if !ok {
			return evaluator.NewBool(false), nil
		}
		return evaluator.NewBool(containsSubstring(in_.Value, valStr.Value)), nil

	case evaluator.A0List:
		// Deep element membership
		for _, item := range in_.Items {
			if evaluator.DeepEqual(item, value) {
				return evaluator.NewBool(true), nil
			}
		}
		return evaluator.NewBool(false), nil

	case evaluator.A0Record:
		// Key existence (value must be string)
		valStr, ok := value.(evaluator.A0String)
		if !ok {
			return evaluator.NewBool(false), nil
		}
		_, found := in_.Get(valStr.Value)
		return evaluator.NewBool(found), nil
	}

	return evaluator.NewBool(false), nil
}

func containsSubstring(s, substr string) bool {
	return len(substr) == 0 || len(s) >= len(substr) && searchString(s, substr)
}

func searchString(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

// and { a, b } → bool
func stdlibAnd(args *evaluator.A0Record) (evaluator.A0Value, error) {
	a, _ := args.Get("a")
	b, _ := args.Get("b")
	if a == nil {
		a = evaluator.NewNull()
	}
	if b == nil {
		b = evaluator.NewNull()
	}
	return evaluator.NewBool(evaluator.Truthiness(a) && evaluator.Truthiness(b)), nil
}

// or { a, b } → bool
func stdlibOr(args *evaluator.A0Record) (evaluator.A0Value, error) {
	a, _ := args.Get("a")
	b, _ := args.Get("b")
	if a == nil {
		a = evaluator.NewNull()
	}
	if b == nil {
		b = evaluator.NewNull()
	}
	return evaluator.NewBool(evaluator.Truthiness(a) || evaluator.Truthiness(b)), nil
}

// coalesce { in, default } → any
// Returns `in` if not null (strict null-check, NOT truthiness), else `default`.
func stdlibCoalesce(args *evaluator.A0Record) (evaluator.A0Value, error) {
	input, _ := args.Get("in")
	fallback, _ := args.Get("default")
	if fallback == nil {
		fallback = evaluator.NewNull()
	}
	if input == nil {
		return fallback, nil
	}
	if _, isNull := input.(evaluator.A0Null); isNull {
		return fallback, nil
	}
	return input, nil
}

// typeof { in } → string
func stdlibTypeof(args *evaluator.A0Record) (evaluator.A0Value, error) {
	input, _ := args.Get("in")
	if input == nil {
		return evaluator.NewString("null"), nil
	}
	switch input.(type) {
	case evaluator.A0Null:
		return evaluator.NewString("null"), nil
	case evaluator.A0Bool:
		return evaluator.NewString("boolean"), nil
	case evaluator.A0Number:
		return evaluator.NewString("number"), nil
	case evaluator.A0String:
		return evaluator.NewString("string"), nil
	case evaluator.A0List:
		return evaluator.NewString("list"), nil
	case evaluator.A0Record:
		return evaluator.NewString("record"), nil
	}
	return evaluator.NewString("null"), nil
}
